### DISCUSSION TITLE
Folding Headers

### DISCUSSION THREAD

**Discussion by shawnh-git** [vkbo/novelWriter#872] [Ideas]
Firstly I want to say I love this program, it's almost exactly what I want a writing program to be. Thank you!

 A feature I would love is folding headers. Is there any chance they could be added? I've become addicted to them through vim and emacs org mode. With folding headers I could make a research/notes/ideas file for the novel like this:

```
# Title
## Ideas
Some Ideas

## Research
### Castles in Europe
Info on castles
### Stew
Delicious stew info

## Characters
### Louise
Lou info
### Bob
This is Robert

## Outline
### Beginning
The start
### Middle
The middle
### End
The ending
```
But folded down would be:
```
# Title
## Ideas
## Research
## Characters
## Outline
```
It would be handy in any case not just for that usage. I would love to do everything in novelwriter but at the moment I make an emacs org file similar to the above and copy and paste sections into novelwriter when I'm ready to actually get stuck in, purely because emacs does folding.

**Comment by vkbo** [comment]
Hi, yes, this has been suggested as a feature in #513.

It is possible to do this. My main concern is that since novelWriter is written in Python, layering features on top of the text editor may slow down the editor too much. The main reason the app works fairly well with Python is that I've mostly left the editor itself alone.

Another limitation here is that the content of the editor is saved as-is to the file system. There is currently no way to save the visibility state of headings to the file format, so every document would start with the headers expanded on each open.

I plan to look into this at some point, but currently there are other features higher on my list. I'm not a huge fan of this feature either. It's fine for a code editor, but for text editors it's more common to opt for an outline consisting of a list of headers that you can navigate from.

**Comment by shawnh-git** [comment]
Thanks for the reply. That's fair, to be honest maybe I should stop trying to force my weird workflow on every program I come across!

Keep up the good work!

**Reply by vkbo** [reply]
Thanks!

I mean, I see a clear value in having some way to get such an overview of a document. Especially project notes, which tend to have many headings. This is especially the case at my end for those notes outlining the flow of the main plot.

Basically, what I propose in #513, is to have a way to show this:
```
# Title
## Ideas
## Research
## Characters
## Outline
```
in a box instead, so that you can navigate to each section by simply clicking on the header.

Firstly, it requires no modifications to the editor itself. The position of each header is already captured and stored in the index each time you save a document, so it is also a trivial matter to generate such a box. I'd like to do what VSCode does and put it as an expandable box under the project tree.

I think this is a decent compromise between your proposal (and #513) and keeping the layout snappy and clean.

Whenever I have the time, I'm working on a C++ app that is nearly identical to novelWriter, but is a rich text editor at the core. There I have much more flexibility in creating such features. It's just a side project, and currently novelWriter takes up most of the time I have for such projects, but at some point it may be an alternative that incorporates more of the ideas and experiences that have come from making novelWriter.

I fully understand the desire to have the best features from code editors incorporated into a writing app. It is partially why I made novelWriter in the first place.

**Reply by shawnh-git** [reply]
> Basically, what I propose in #513, is to have a way to show this:
> 
> ```
> # Title
> ## Ideas
> ## Research
> ## Characters
> ## Outline
> ```
> 
> in a box instead, so that you can navigate to each section by simply clicking on the header.

That's a great idea, I think ghostwriter does a similar thing but in a strange transparent floating window. I would love a way to sensibly keep research and notes for each story contained within a novelwriter project but at the moment it would require massive amounts of scrolling/searching. It's certainly still possible, but your idea would make it so much easier!

**Reply by vkbo** [reply]
Yes, GhostWriter has such a box. So does Zim. Although the transparent floating boxes are gone in GhostWriter 2.0 and have been replaced by a side panel. I did the Norwegian translations for that version, but I don't think it's trickled down to the distro repos yet.

In any case, the idea is to split up the project notes into separate files. Sure, my main plot note is easily 10k words or more, but for most things else, I have a document per character, location, etc. That makes it easier to add references to them in the novel text when you start writing.
